src/clients/Deluge.ts

```diff
@@ -0,0 +1,236 @@
+import { getRuntimeConfig } from "../runtimeConfig.js";
+import { Searchee } from "../searchee.js";
+import { TorrentClient } from "./TorrentClient.js";
+
+export default class Deluge implements TorrentClient {
+	private msgId = 0;
+	private loggedIn = false;
+	private delugeCookie = "";
+	private delugeWebUrl: URL;
+	private delugeLabel: string;
+	private enLabeled: Promise<boolean>;
+
+	constructor() {
+		this.delugeWebUrl = new URL(`${getRuntimeConfig().delugeWebUrl}`);
+		this.delugeLabel = getRuntimeConfig().dataCategory;
+	}
+
+	/**
+	 * validates the login and host for deluge webui
+	 */
+	async validateConfig(): Promise<void> {
+		const url = new URL(this.delugeWebUrl);
+		if (url.username && !url.password) {
+			throw new CrossSeedError(
+				"you need to define a password in the delugeWebUrl. (eg: http://:<PASSWORD>@localhost:8112)"
+			);
+		}
+		const result = await this.authenticate();
+		if (!result) {
+			logger.verbose({
+				label: Label.DELUGE,
+				message: "failed to authenticate the deluge daemon.",
+			});
+			throw new CrossSeedError(
+				`failed to authenticate the deluge daemon.`
+			);
+		} else {
+			this.enLabeled = this.labelEnabled();
+		}
+	}
+
+	/**
+	 * connects and authenticates to the webui
+	 */
+	private async authenticate(): Promise<boolean> {
+		if (!this.loggedIn) {
+			let response = await this.call({
+				params: [this.delugeWebUrl.password],
+				method: "auth.login",
+			});
+			if (response === null) {
+				throw new CrossSeedError(
+					`Could not reach deluge: ${this.delugeWebUrl.origin}`
+				);
+			}
+			if (
+				response &&
+				response.headers &&
+				response["data"]["result"] &&
+				response.headers["set-cookie"]
+			) {
+				this.delugeCookie =
+					response.headers["set-cookie"][0].split(";")[0];
+				this.loggedIn = true;
+			}
+		} else {
+			let response = await this.call({
+				params: [this.delugeCookie],
+				method: "auth.check_session",
+			});
+			this.loggedIn = !!response;
+		}
+		return this.loggedIn;
+	}
+
+	/**
+	 * authenticates and sends JSON-RPC calls to deluge
+	 */
+	private async sendCall(method: string, params: any[]) {
+		let response;
+		if (!this.loggedIn) {
+			await this.authenticate();
+			response = await this.call({
+				params,
+				method,
+			});
+		} else
+			response = await this.call({
+				params: params,
+				method,
+			});
+
+		return response && response.data && response.data.error === null
+			? response.data
+			: response.data.error;
+	}
+
+	private async call(body: any): Promise<AxiosResponse | null> {
+		return new Promise((resolve) => {
+			body.id = ++this.msgId;
+			if (this.msgId > 1024) this.msgId = 0;
+			axios
+				.post(
+					this.delugeWebUrl.origin + this.delugeWebUrl.pathname,
+					body,
+					{
+						headers: {
+							"Content-Type": "application/json",
+							Cookie: this.delugeCookie,
+						},
+					}
+				)
+				.then((data) => {
+					resolve(data);
+				})
+				.catch(() => resolve(null));
```

Comment:
I'd rather rethrow than resolve with `null` and lose the error object here. 

lmk if you want some more detail on how Promise error handling works, also [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise) are some high quality docs on promises 🙂